tests/melee_test.cpp

```diff
@@ -0,0 +1,220 @@
+        INFO( full_attack_details( dude ) );
+        check_near( prob, 0.95f, 0.05f );
+    }
+}
+
+TEST_CASE("Character attacking a manhack", "[melee]") {
+    monster manhack( mtype_id( "mon_manhack" ) );
+    INFO( "Manhack has get_dodge() == " + std::to_string( manhack.get_dodge() ) );
+
+    SECTION("8/8/8/8, no skills, unarmed") {
+        standard_npc dude( "TestCharacter", {}, 0, 8, 8, 8, 8 );
+        float prob = brute_probability( dude, manhack, num_iters );
+        INFO( "Has get_hit() == " + std::to_string( dude.get_hit() ) );
+        check_near( prob, 0.15f, 0.05f );
+    }
+
+    SECTION("8/8/8/8, 3 all skills, two-by-four") {
+        standard_npc dude( "TestCharacter", {}, 3, 8, 8, 8, 8 );
+        dude.weapon = item( "2x4" );
+        float prob = brute_probability( dude, manhack, num_iters );
+        INFO( "Has get_hit() == " + std::to_string( dude.get_hit() ) );
+        check_near( prob, 0.4f, 0.05f );
+    }
+
+    SECTION("10/10/10/10, 8 all skills, katana") {
+        standard_npc dude( "TestCharacter", {}, 8, 10, 10, 10, 10 );
+        dude.weapon = item( "katana" );
+        float prob = brute_probability( dude, manhack, num_iters );
+        INFO( "Has get_hit() == " + std::to_string( dude.get_hit() ) );
+        check_near( prob, 0.7f, 0.05f );
+    }
+}
+
+TEST_CASE("Zombie attacking a character", "[melee]") {
+    monster zed( mtype_id( "mon_zombie" ) );
+    INFO( "Zombie has get_hit() == " + std::to_string( zed.get_hit() ) );
+
+    SECTION("8/8/8/8, no skills, unencumbered") {
+        standard_npc dude( "TestCharacter", {}, 0, 8, 8, 8, 8 );
+        float prob = brute_probability( zed, dude, num_iters );
+        INFO( "Has get_dodge() == " + std::to_string( dude.get_dodge() ) );
+        THEN( "Character has no significant dodge bonus or penalty" ) {
+            REQUIRE( dude.get_dodge_bonus() < 0.5f );
+            REQUIRE( dude.get_dodge_bonus() > -0.5f );
+        }
+
+        THEN( "Character's dodge skill is roughly equal to zombie's attack skill" ) {
+            REQUIRE( dude.get_dodge() < zed.get_hit() + 0.5f );
+            REQUIRE( dude.get_dodge() > zed.get_hit() - 0.5f );
+        }
+
+        check_near( prob, 0.5f, 0.05f );
+    }
+
+    SECTION("10/10/10/10, 3 all skills, good cotton armor") {
+        standard_npc dude( "TestCharacter", { "hoodie", "jeans", "long_underpants", "long_undertop", "longshirt" }, 3, 10, 10, 10, 10 );
+        float prob = brute_probability( zed, dude, num_iters );
+        INFO( "Has get_dodge() == " + std::to_string( dude.get_dodge() ) );
+        check_near( prob, 0.2f, 0.05f );
+    }
+
+    SECTION("10/10/10/10, 8 all skills, survivor suit") {
+        standard_npc dude( "TestCharacter", { "survivor_suit" }, 8, 10, 10, 10, 10 );
+        float prob = brute_probability( zed, dude, num_iters );
+        INFO( "Has get_dodge() == " + std::to_string( dude.get_dodge() ) );
+        check_near( prob, 0.025f, 0.0125f );
```

Comment:
When the hit chance is very low and very high, the tolerance can be lower because the variance gets lower.
For example, changing from `to_hit - dodge` from 0 to 1 will result in ~10% change in hit ratio, but changing it from 10 to 11 will only change it by something like 2%.
I'd calculate it from the expected value, but that would lead to "ugly" numbers.